File: public/javascripts/playerEngine.js
define(["lib/monet", "report", "command"], function (monet, _report, _command) {

class PlayerEngine {
    constructor(commLink, factions) {
        this.commLink = commLink;
        this.factions = factions;
    }

    getFaction(factionId) {
        return _.findWhere(this.factions, {id: factionId});
    }

    getNewReports() {
        return this.commLink.getNewReports();
    }

    getLatestFleetStatuses() {
        return _.chain(this.commLink.getAllReports())
            .filter(report => isStatusReport(report) || isDepartureReport(report)) // status, and any report that could remove a fleet from a system
            .groupBy(report => report.fleetId)
            .map(reports => _.max(reports, report => report.time)) // get the latest report
            .filter(report => isStatusReport(report)) // now, only return the status reports
            .value();
    }


    /**
     * Get the last report received from a fleet
     * @return A report object
     */
    getLatestReport(fleetId) {
        return this._genericGetFleetReport(fleetId, report => true);
    }

    /**
     * Get the last status report received from a fleet
     */
    getLatestStatusReport(fleetId) {
        return this._genericGetFleetReport(fleetId, isStatusReport);
    }

    /**
     * Get the last known star system that a given fleet was in
     * @param fleetId
     * @return An option of a system. None if fleet is in transit
     */
    getSystem(fleetId) {
        var latestReport = this.getLatestReport(fleetId);
        if (isDepartureReport(latestReport)) {
            return monet.None();
        } else {
            return monet.Some(latestReport.origin);
        }
    }

    /**
     * The latest reports about each fleet in a given system
     * @param systemId
     * @returns {*} A list of reports
     */
    getFleetsInSystem(systemId) {
        return _.chain(this.getLatestFleetStatuses())
            .filter(report => !isDepartureReport(report))
            .filter(report => report.origin.id == systemId)
            .value();
    }

    getStarSystems() {
        return this.commLink.getStarSystems();
    }

    // Actions
    createFleet() {
        this.commLink.createFleet();
    }

    sendFleetToSystem(fleetId, systemId) {
        this.commLink.sendCommand(_command.GoTo(fleetId, systemId));
    }

    /**
     * Send all the fleets currently known to be in a system to another system
     * @param fromSystemId
     * @param systemId
     */
    sendFleetsInSystem(fromSystemId, systemId) {
        var fleetReports = this.getFleetsInSystem(fromSystemId);
        fleetReports.forEach(fleetReport => this.commLink.sendCommand(_command.GoTo(fleetReport.fleetId, systemId)));
    }

    addReportReceivedCallback(callback) {
        this.commLink.addReportReceivedCallback(callback);
    }

    /**
     * Helper method to get the latest report from a fleet
     */
    _genericGetFleetReport(fleetId, reportFilter) {
        return _.chain(this.commLink.getAllReports())
            .filter(reportFilter)
            .filter(report => report.fleetId == fleetId)
            .max(report => report.time)
            .value();
    }
}

function isStatusReport(report) {
    return report.type == _report.ReportTypes.Status;
}

function isDepartureReport(report) {
    return report.type == _report.ReportTypes.Departure;
}

return {
    PlayerEngine: PlayerEngine
}
});

File: src/client/PlayerEngine.scala
package client

import java.util.UUID

import core._
import networking.InGameCommunication

/**
  * Created by dan-s on 26/05/2016.
  */

// knows only about reports, and maybe predicted positions for fleets in transit
class PlayerEngine(commLink: InGameCommunication) {
  def getNewReports = commLink.getNewReports

  // The most recent status report for each fleet the player controls
  def latestFleetStatuses: Seq[StatusReport] =
  // TODO handle fleet destruction - should not get a status report?
    commLink.getAllReports
      .collect({
        case report: StatusReport => report
      })
      .groupBy(_.fleetId).values  // group the reports by fleet
      .map(reports =>
        // For each list of reports for a single fleet, find the latest one
        reports.maxBy(_.time)
      )
      .toList

  /**
    * Get the last report received from a fleet
    */
  def getLatestReport(fleetId: UUID): Report = {
    commLink.getAllReports.filter(_.fleetId == fleetId).maxBy(_.time)
  }
  /**
    * Get the last status report received from a fleet
    */
  def getLatestStatusReport(fleetId: UUID): StatusReport = {
    latestFleetStatuses.find(_.fleetId == fleetId).getOrElse(throw new Exception(s"No report for fleet $fleetId"))
  }

  /**
    * Get the last known star system for a fleet
    * @return Some if in system, None if fleet is in transit
    */
  def getSystem(fleetId: UUID): Option[StarSystem] = {
    val latestReport = getLatestReport(fleetId)
    latestReport match {
      case r: DepartureReport => None
      case _ => Some(latestReport.origin)
    }
  }

  // todo this is wrong
  def getFleetsInSystem(systemId: UUID): Seq[Report] = {
    latestFleetStatuses.filterNot(_.isInstanceOf[DepartureReport])
  }

  var systems: Seq[StarSystem] = commLink.getStarSystems


  def addFleet(): Unit = {
    commLink.createFleet()
  }

  def goTo(systemId: UUID, fleetId: UUID): Unit = {
    commLink.sendAction(GoTo(systemId, fleetId))
  }
}
